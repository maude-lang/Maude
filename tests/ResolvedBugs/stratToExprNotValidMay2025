#!/bin/sh

MAUDE_LIB=$srcdir/../../src/Main
export MAUDE_LIB

../../src/Main/maude \
  < $srcdir/stratToExprNotValidMay2025.maude -no-banner -no-advise \
  > stratToExprNotValidMay2025.out 2>&1

diff $srcdir/stratToExprNotValidMay2025.expected stratToExprNotValidMay2025.out > /dev/null 2>&1
