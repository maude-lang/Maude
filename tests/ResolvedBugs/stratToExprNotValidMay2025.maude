***
***	Test for strat to expr mappings that do not satisfy the requirements:
*** either the strategy is not well-formed (unbound variables), the are
*** repeated variables in the left-hand side, or non-variable terms.
***
*** Maude should show an warning and invalidate the view, not an internal error.
***

sth F is
	sort Elt .
	op 0 : -> Elt .
	strat f : Elt @ Elt .
endsth

sth G is
	sort Elt .
	strat g : Elt Elt @ Elt .
endsth

view NatF1 from F to NAT is
	sort Elt to Nat .
	strat f(N:Elt) to expr match M:Nat s.t. M:Nat = J:Nat .
endv

view NatF2 from F to NAT is
	sort Elt to Nat .
	strat f(0) to expr idle .
endv

view NatG from G to NAT is
	sort Elt to Nat .
	strat g(N:Elt, N:Elt) to expr idle .
endv

smod PARAM-F{X :: F} is
endsm

smod PARAM-G{X :: F} is
endsm

smod MAIN is
	protecting PARAM-F{NatF1} .
	protecting PARAM-F{NatF2} .
	protecting PARAM-G{NatG} .
endsm
