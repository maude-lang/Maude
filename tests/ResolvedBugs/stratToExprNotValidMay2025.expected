Warning: <standard input>, line 22 (view NatF1): variable J:Nat is used before
    it is bound in condition of test strategy.
Warning: <standard input>, line 27 (view NatF2): lhs of strategy mapping has
    non-variable argument 0.
Warning: <standard input>, line 32 (view NatG): using the same variable base
    name N for two left hand side variables in an strategy mapping is not
    allowed.
Warning: <standard input>, line 42 (smod MAIN): unusable view NatF1.
Warning: <standard input>, line 43 (smod MAIN): unusable view NatF2.
Warning: <standard input>, line 44 (smod MAIN): unusable view NatG.
Warning: <standard input>, line 41 (smod MAIN): this module contains one or
    more errors that could not be patched up and thus it cannot be used or
    imported.
Bye.
